### ISSUE TITLE
`@__FILE__` not available on workers

### ISSUE THREAD

**Issue by artas360** [JuliaLang/julia#17500]
Hi all,
it seems that the `@__FILE__` macro doesn't not return the expected result when evaluated by workers.
To illustrate this issue you can create a `test.jl` script containing the following code:

``` julia
@everywhere f = @__FILE__;
println(pmap(x->f, workers()));
```

Which, when running `julia test.jl` prints `Any["/tmp/test.jl"]` as expected. However with `julia -p2` the result is `Any["",""]`, which leads to some issues when trying to reliably access files across workers.

The tests were ran with:

```
Julia Version 0.4.6
Commit 2e358ce (2016-06-19 17:16 UTC)
```

and

```
Julia Version 0.5.0-dev+5504
Commit c7ca319 (2016-07-19 12:45 UTC)
```

**Comment by amitmurthy** [comment]
A workaround for this is 
```
foo = @__FILE__
@eval @everywhere f = $foo
```

The expression `f=@__FILE__` on workers cannot really return a pathname as `@__FILE__` on the worker does not not point to `test.jl`, in fact it is not even loaded on the workers.

I am inclined to close this issue as a wont-fix

**Comment by tkelman** [comment]
could at least use a note in `@__FILE__`'s documentation

**Comment by vtjnash** [comment]
could also mention `Base.source_path()` too

**Comment by artas360** [comment]
@amitmurthy Thanks a lot for the work around! :) I'll be using it.

**Comment by vtjnash** [comment]
fixed by #21746

**Comment by tkelman** [comment]
likely to regress if it doesn't get tested
